Dropping DB

ajmejia opened the pull request:
# Implementing metadata caching without DB

These changes are focused on keeping track of relevant metadata (`mjd`, `imagetyp`, `camera`, `expnum`, `exptime`, etc) to be able to match analog frames and master calibration frames quickly, and to keep track of the reduction `stage`, `status` and `quality` as the frames are being reduced.

The following changes are introduced:

- Implemented metadata caching using HDF5 in place of a SQLite DB
- Implemented low-level functions `_*` for general metadata manipulation
- Implemented high-level functions that are meant to be run in the following sequence:
  - `extract_metadata`: to extract header information from frames stored locally
  - `add_metadata`: add metadata to the HDF5 store
  - `del_metadata`: remove HDF5 store or a given `mjd` within it
  - `get_metadata`: retrieve metadata from the store
  - `get_analog_groups`: group metadata into analogs for combining into master frames
  - `match_master_metadata`: match master calibration frames with target frames
  - `put_reduction_stage`: update the reduction stage of a particular frame
- The matching functions follow the criteria:
  - Analogs are based on their `mjd`, `camera`, `imagetyp` and `exptime`
  - Calibration-target frames are based on their `mjd`, `camera`, `imagetyp`, `exptime`, `quality`, `stage` and arc lamp
- The module `lvmdrp.utils.database` is replaced by `lvmdrp.utils.metadata`

Some bug fixes were also introduced to the `runMethod` module, but those were only meant for testing the `metadata` module.

havok2063 commented on python/lvmdrp/utils/metadata.py:
```diff
@@ -0,0 +1,828 @@
+    h5py.Group
+        the metadata store for the given observatory
+    """
+    metadata_path = os.path.join(access.base_dir, "metadata.hdf5")
```
The base dir in sdss_access is the root directory of the SAS.  LVM can't write to this directory.  Instead, this file should live inside LVM's reduction directory at $LVM_SPECTRO_REDUX/{drpver}/.  

Depending on the size of this file, and how it grows over time, it may make more sense to split this file up by MJD, and store it inside the the [tileid]/[mjd] subdirectories.  Would that affect how you use this file?

havok2063 replied:
Ok, great.  I'll test this as well.

havok2063 commented on python/lvmdrp/utils/metadata.py:
```diff
@@ -0,0 +1,828 @@
+        header = fits.getheader(frame_path, ext=0)
+        new_metadata[i] = [
+            "n" if header.get("OBSERVAT") != "LCO" else "s",
+            header.get("IMAGETYP"),
```
The `TILEID` header keyword should be included here.   If not already in there, can you also include the spectrograph ID and the file name as columns?  Just the file name is fine, not the full path.

havok2063 replied:
Ok sounds good.

havok2063 commented on python/lvmdrp/utils/metadata.py:
```diff
@@ -0,0 +1,828 @@
+            header.get("ARGON", "OFF") == "ON",
+            header.get("LDLS", "OFF") == "ON",
+            header.get("QUARTZ", "OFF") == "ON",
+            header.get("QUALITY", QualityFlag(0)),
```
The `QUALITY` flag in the raw frame header only refers to whether the raw exposure was a test exposure, is bad, or is good for use.  This is different than the reduction quality status of the exposure, which I think is was the `QualityFlag` refers to.  If you're planning on updating this column throughout the reduction then it's probably better this is a brand new column, e.g. `DRPQUAL`, which needs to get updated and written out to the reduced file headers.   I think we want both columns in this table.

ajmejia replied:
I added a new bitmask class `lvmdrp.utils.bitmask.RawFrameQuality`:

```python
class RawFrameQuality(BaseBitmask):
    GOOD = auto()  # bit whether a raw frame is good for reduction
    TEST = auto()  # bit whether a raw frame is for instrument testing purposes
    BAD = auto()  # bit whether a raw frame is bad for reduction
```

This will be stored in the metadata tables as the `drpqual` column.

havok2063 replied:
I think the `RawFrameQuality` should map to the `QUALITY` header, but I'm not sure we want to overwrite the original keyword.  The `QualityFlag` is what should get mapped to the new DRPQUAL header keyword.

ajmejia replied:
Yes, that's how those bitmask flags will map. I don't think we should be overwriting header keywords coming from the original (raw) frames.

havok2063 commented on python/lvmdrp/utils/metadata.py:
```diff
@@ -0,0 +1,828 @@
+    return analogs
+
+
+def match_master_metadata(
```
Does this only return the relevant master biases and darks to use for a reduction?  Or it does also provide the closest matching reduced arc and flat to use for reducing a science frame?

havok2063 replied:
what happens if there are no matching relevant calibration frames?

ajmejia replied:
yes, this function will return relevant master calibration frames: bias, dark, fiberflat and arcs. The general structure of the output is a dictionary with calibration frame `imagetyp`s for keys and values being the relevant master calibration metadata (as a pandas series).

If no calibration frames are not found, the values will be `None`.

havok2063 replied:
Ok, great.  I will test this out.  So in principle reducing 60041 will find the right cals from 60039-40?

havok2063 reviewed (changes requested):
I think this mostly looks good.  I've requested some changes.  There might be some additional changes to get it working with the main pipeline code but we can tweak those later as needed.

ajmejia commented:
The functionality of the `metadata` module is mostly the same after implementing these requests. In summary:

- The metadata file structure has changed to store the metadata in the reduction paths.
- There will be metadata stores for raw frames for each `tileid` and `mjd`.
- There will be one metadata store for master calibration frames at the root of a reduction version.
- To retain the possibility to match multiple calibration frames across `tileid`s and `mjd`s, I implemented store matching with wildcards.
- I dropped the `observatory` parameter/column in all functions/tables as that information is already in the reduction paths. In the tables the column `hemi` was added to be able to match frame paths using the `path.expand` method.

havok2063 reviewed (approved):
I approve this.  I'll merge this in and work on getting it working with my code.  I might make some changes to the code to facilitate that but I will let you know what they are.

havok2063 commented:
This branch this work was in should be deleted.
